homeassistant/components/upcloud/config_flow.py

```diff
@@ -0,0 +1,112 @@
+"""Config flow for UpCloud."""
+
+import logging
+
+import requests.exceptions
+import upcloud_api
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_SCAN_INTERVAL, CONF_USERNAME
+from homeassistant.core import callback
+
+# pylint: disable=unused-import # for DOMAIN https://github.com/PyCQA/pylint/issues/3202
+from .const import DEFAULT_SCAN_INTERVAL, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class UpCloudConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """UpCloud config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    username: str
+    password: str
+
+    async def async_step_user(self, user_input=None):
+        """Handle user initiated flow."""
+        if user_input is None:
```

Comment:
Our standard is to have this check inverted and just call show form in end of the step body. Then we can also inline the form helper method `_async_show_form` in the user step.

Reply:
Not sure I follow you correctly, but inverting the check leads to a largish block of indented code containing the connection/authentication checks which are dependent on user input. We seem to have a standard of guarding against such blocks with an early if+return, which is what is being done here.

Reply:
No, in the config flow steps our standard is the opposite like I described.

Reply:
Here's our scaffold template:
https://github.com/home-assistant/core/blob/d16c04a8e06e023626ae74e465c0044a20904bdd/script/scaffold/templates/config_flow/integration/config_flow.py#L65-L83

Reply:
Yes, I understand what you mean. But as I described, I think we do have a more general standard of returning early and avoiding large indented blocks, and that scaffold template also goes against that (and should be fixed if you ask me). This has come up in many reviews, and I believe you also have (rightfully) also asked me to do that somewhere (won't go searching "evidence" right now though). But there's [one example](https://github.com/home-assistant/core/pull/37941#discussion_r500153901) in this very review where Paulus did.

Reply:
Yes, normally it's good to return early. But not in this case. Here making it DRY is better. This is our current standard and we should follow that.

Reply:
Following the earlier suggestion is not going to do anything wrt DRY. It would just result in a change to indent level of one code block, and inline one function in that block. No repetition related changes.

Reply:
We're calling show form twice instead of once. I'm not going to discuss our standard more in this PR. We should follow our scaffold template.